cdap-storage-spi/src/test/java/co/cask/cdap/spi/data/StructuredTableConcurrencyTest.java

```diff
@@ -0,0 +1,364 @@
+    Long value = 777L;
+    Map<Long, Boolean> updateMap = new ConcurrentHashMap<>();
+    runConcurrentOperation("concurrent-compare-and-swap", numThreads, () ->
+      TransactionRunners.run(transactionRunner, context -> {
+        StructuredTable table = context.getTable(TEST_TABLE);
+        boolean success = table.compareAndSwap(keys, Fields.longField(LONG_COL, null),
+                                               Fields.longField(LONG_COL, value));
+        updateMap.put(Thread.currentThread().getId(), success);
+      }));
+
+    Assert.assertEquals(numThreads, updateMap.size());
+    // Assert only one compare and swap was successful
+    int numSuccess = 0;
+    for (Map.Entry<Long, Boolean> entry : updateMap.entrySet()) {
+      if (entry.getValue()) {
+        numSuccess++;
+      }
+    }
+    Assert.assertEquals(1, numSuccess);
+    // Assert value
+    Optional<StructuredRow> row = TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable table = context.getTable(TEST_TABLE);
+      return table.read(keys, Collections.singleton(LONG_COL));
+    });
+    Assert.assertTrue(row.isPresent());
+    Assert.assertEquals(value, row.get().getLong(LONG_COL));
+  }
+
+  @Test
+  public void testConcurrentDelete() throws Exception {
+    // Write one row
+    Field<Integer> key = Fields.intField(KEY, 1);
+    String strVal = "str1";
+    Long longVal = 350L;
+    List<Field<?>> fields = Arrays.asList(key,
+                                          Fields.stringField(STRING_COL, strVal),
+                                          Fields.longField(LONG_COL, longVal));
+    TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable table = context.getTable(TEST_TABLE);
+      table.upsert(fields);
+    });
+
+    // Assert write
+    Optional<StructuredRow> row = TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable table = context.getTable(TEST_TABLE);
+      return table.read(Collections.singleton(key));
+    });
+    Assert.assertTrue(row.isPresent());
+    Assert.assertEquals(longVal, row.get().getLong(LONG_COL));
+
+    // delete it concurrently
+    int numThreads = 15;
+    runConcurrentOperation("concurrent-read", numThreads, () ->
+      TransactionRunners.run(transactionRunner, context -> {
+        StructuredTable table = context.getTable(TEST_TABLE);
+        table.delete(Collections.singleton(key));
+      }));
+
+    // Verify delete
+    row = TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable table = context.getTable(TEST_TABLE);
+      return table.read(Collections.singleton(key));
+    });
+    Assert.assertFalse(row.isPresent());
+  }
+
+  private void runConcurrentOperation(String name, int numThreads, Runnable runnable) throws Exception {
+    ExecutorService executorService = Executors.newFixedThreadPool(numThreads);
+    CountDownLatch startLatch = new CountDownLatch(1);
```

Comment:
What is the usage of the `startLatch`? The `doneLatch` is already recording each thread operations is successful or not, right?

Reply:
The `startLatch` is used to make sure that all the threads start at the same time to maximize the chances of race condition

Reply:
👍 got it